src/v/kafka/tests/fetch_test.cc

```diff
@@ -281,3 +281,242 @@ FIXTURE_TEST(fetch_one, redpanda_thread_fixture) {
+
+    for (int i = 0; i < 6; ++i) {
+        auto ntp = make_default_ntp(topic, model::partition_id(i));
+        auto shard = app.shard_table.local().shard_for(ntp);
+        tests::cooperative_spin_wait_with_timeout(10s, [this, shard, ntp] {
+            return app.partition_manager.invoke_on(
+              *shard, [ntp](cluster::partition_manager& mgr) {
+                  auto partition = mgr.get(ntp);
+                  return partition
+                         && partition->committed_offset() >= model::offset(0);
+              });
+        }).get();
+    }
+
+    kafka::fetch_request req;
+    req.max_bytes = std::numeric_limits<int32_t>::max();
+    req.min_bytes = 1;
+    req.max_wait_time = std::chrono::milliseconds(3000);
+    req.topics = {{
+      .name = topic,
+      .partitions = {},
+    }};
+    for (int i = 0; i < 6; ++i) {
+        kafka::fetch_request::partition p;
+        p.id = model::partition_id(i);
+        p.log_start_offset = offset;
+        p.fetch_offset = offset;
+        p.partition_max_bytes = std::numeric_limits<int32_t>::max();
+        req.topics[0].partitions.push_back(p);
+    }
+    auto client = make_kafka_client().get0();
+    client.connect().get();
+    auto fresp = client.dispatch(req, kafka::api_version(4));
+
+    for (int i = 0; i < 6; ++i) {
+        model::partition_id partition_id(i);
+        auto ntp = make_default_ntp(topic, partition_id);
+        auto shard = app.shard_table.local().shard_for(ntp);
+        auto r = app.partition_manager
+                   .invoke_on(
+                     *shard,
+                     [ntp](cluster::partition_manager& mgr) {
+                         auto partition = mgr.get(ntp);
+                         auto batches = storage::test::make_random_batches(
+                           model::offset(0), 5);
+                         auto rdr = model::make_memory_record_batch_reader(
+                           std::move(batches));
+                         return partition->replicate(
+                           std::move(rdr),
+                           raft::replicate_options(
+                             raft::consistency_level::quorum_ack));
+                     })
+                   .get0();
+    }
+    auto resp = fresp.get0();
+    client.stop().then([&client] { client.shutdown(); }).get();
+
+    BOOST_REQUIRE_EQUAL(resp.partitions.size(), 1);
+    BOOST_REQUIRE_EQUAL(resp.partitions[0].name, topic());
+    BOOST_REQUIRE_EQUAL(resp.partitions[0].responses.size(), 6);
+    size_t total_size = 0;
+    for (int i = 0; i < 6; ++i) {
+        BOOST_REQUIRE_EQUAL(
+          resp.partitions[0].responses[i].error, kafka::error_code::none);
+        BOOST_REQUIRE_EQUAL(
+          resp.partitions[0].responses[i].id, model::partition_id(i));
+        BOOST_REQUIRE(resp.partitions[0].responses[i].record_set);
+
+        total_size += resp.partitions[0].responses[i].record_set->size_bytes();
+    }
+    BOOST_REQUIRE_GT(total_size, 0);
```

Comment:
Maybe we should also verify that all of the data is read. maybe by checking for equivalency between send and received records or for now by verifying the total size read is equal to the bytes sent.

Reply:
We allow `fetch_request` to return after reaching min bytes threshold. We can not expect it to return all the records.